[MRG+1] Fixes #9199 - Fitting a NearestNeighbors model fails with sparse input and a callable as metric

tttthomasssss opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/scikit-learn/scikit-learn/blob/master/CONTRIBUTING.md#Contributing-Pull-Requests
-->
#### Reference Issue
<!-- Example: Fixes #1234 -->

Fixes #9199 

#### What does this implement/fix? Explain your changes.

The fix simply extends the current sparse input incompatibility check with a callable check.

#### Any other comments?

A test is included in `test_neighbors.py`

<!--
Please be aware that we are a loose team of volunteers so patience is
necessary; assistance handling other issues is very welcome. We value
all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.
For more information, see our FAQ on this topic:
http://scikit-learn.org/dev/faq.html#why-is-my-pull-request-not-getting-any-attention.

Thanks for contributing!
-->

tttthomasssss committed:
proposed fix (incl test) for #9199 - Fitting a NearestNeighbors model fails with sparse input and a callable as metric

tttthomasssss committed:
Merge branch 'master' of https://github.com/scikit-learn/scikit-learn

jnothman commented:
Unfortunately the earliest supported version of scipy does not have count_nonzero. Is `np.diff(X.indptr)` sufficient?

Also, you should use PEP8 to pass our tests.

tttthomasssss committed:
fixed scipy compatibility issue in test case and pep8 issues

tttthomasssss commented:
I've just modified my PR - apologies for the `pep8` mess...

jnothman commented:
Tests are failing. Can you please make the title of the PR more descriptive as well?

tttthomasssss committed:
Merge branch 'master' of https://github.com/scikit-learn/scikit-learn

tttthomasssss commented:
@jnothman - do you know which ones are affected by my change? I've removed my fix and the same tests (+ the one I added) are failing in comparison to with my fix (- the one I added).

jnothman commented on sklearn/neighbors/tests/test_neighbors.py:
```diff
@@ -1255,6 +1255,35 @@ def test_dtype_convert():
 
+def test_sparse_metric_callable():
+    def sparse_metric(x, y):  # Metric accepting sparse matrix input (only)
+
```
remove blank line, or put comment here.

jnothman reviewed (commented):
LGTM

jnothman commented on sklearn/neighbors/tests/test_neighbors.py:
```diff
@@ -1255,6 +1255,35 @@ def test_dtype_convert():
+def test_sparse_metric_callable():
+    def sparse_metric(x, y):  # Metric accepting sparse matrix input (only)
+
+        return min(x.nnz, y.nnz) / max(x.nnz, y.nnz)
```
Sorry, I have to take back the LGTM. :\ This is apparently producing different results on old scipy, and tests are failing. Perhaps just assert isspmatrix(x) and return some more ordinary function of the inputs.

tttthomasssss committed:
updated tests, sparse_metric now simply returns the dot product of the inputs, provided they are sparse

tttthomasssss commented:
OK, finally managed to get back to this, the metric now returns the dot product of the inputs

jnothman reviewed (commented):
Thanks. LGTM. Properly this time :)

TomDLT commented:
LGTM, thanks for the fix @tttthomasssss

jnothman commented:
There's no what's new entry. My bad :(

@tttthomasssss, would you be able to open a new PR which adds an entry to the change log at `doc/whats_new/v0.20.rst`. Like the other entries there, please reference this pull request with `:issue:` and credit yourself (and other contributors if applicable) with `:user:`
